src/dwarf_parser.cpp

```diff
@@ -1,419 +1,276 @@
 
 namespace bpftrace {
 
-struct FuncInfo {
-  std::string name;
-  Dwarf_Die die;
-};
+std::atomic<size_t> Dwarf::instance_count = 0;
 
-Dwarf::Dwarf(BPFtrace *bpftrace, const std::string &file_path)
-    : bpftrace_(bpftrace), file_path_(file_path)
+Dwarf::Dwarf(BPFtrace *bpftrace, std::string file_path)
+    : bpftrace_(bpftrace), file_path_(std::move(file_path))
 {
-  callbacks.find_debuginfo = dwfl_standard_find_debuginfo;
-  callbacks.section_address = dwfl_offline_section_address;
-  callbacks.debuginfo_path = NULL;
-  dwfl = dwfl_begin(&callbacks);
-  dwfl_report_offline(dwfl, file_path.c_str(), file_path.c_str(), -1);
-  dwfl_report_end(dwfl, NULL, NULL);
-}
+  if (instance_count++ == 0)
+    lldb::SBDebugger::Initialize();
 
-std::unique_ptr<Dwarf> Dwarf::GetFromBinary(BPFtrace *bpftrace,
-                                            const std::string &file_path)
-{
-  std::unique_ptr<Dwarf> dwarf(new Dwarf(bpftrace, file_path));
-  Dwarf_Addr bias;
-  if (dwfl_nextcu(dwarf->dwfl, NULL, &bias) == NULL)
-    return nullptr;
+  debugger_ = lldb::SBDebugger::Create(false);
 
-  return dwarf;
+  lldb::SBError error;
+  target_ = debugger_.CreateTarget(
+      file_path_.c_str(), nullptr, nullptr, false, error);
+  if (!error.Success() || !target_.IsValid()) {
+    throw error;
+  }
 }
 
 Dwarf::~Dwarf()
 {
-  dwfl_end(dwfl);
+  if (--instance_count == 0)
+    lldb::SBDebugger::Terminate();
 }
 
-static int get_func_die_cb(Dwarf_Die *func_die, void *arg)
+std::unique_ptr<Dwarf> Dwarf::GetFromBinary(BPFtrace *bpftrace,
+                                            std::string file_path)
 {
-  auto *func_info = static_cast<struct FuncInfo *>(arg);
-  if (dwarf_hasattr(func_die, DW_AT_name) &&
-      dwarf_diename(func_die) == func_info->name) {
-    func_info->die = *func_die;
-    return DWARF_CB_ABORT;
+  try {
+    // Can't use std::make_unique because Dwarf's constructor is private
+    return std::unique_ptr<Dwarf>(new Dwarf(bpftrace, std::move(file_path)));
+  } catch (const lldb::SBError &error) {
+    LOG(ERROR) << "Failed to parse DebugInfo: " << error.GetCString();
+    return nullptr;
   }
-  return DWARF_CB_OK;
 }
 
-std::optional<Dwarf_Die> Dwarf::get_func_die(const std::string &function) const
+std::string Dwarf::get_type_name(lldb::SBType type)
 {
-  struct FuncInfo func_info = { .name = function, .die = {} };
-
-  Dwarf_Die *cudie = nullptr;
-  Dwarf_Addr cubias;
-  while ((cudie = dwfl_nextcu(dwfl, cudie, &cubias)) != nullptr) {
-    if (dwarf_getfuncs(cudie, get_func_die_cb, &func_info, 0) > 0)
-      return func_info.die;
+  std::string type_name = type.GetDisplayTypeName() ?: "<anonymous type>";
+
+  /* LLDB removes the struct/union/enum prefix from the type name, but C
+   * requires it! So we first need to peel off pointers/references to reach the
+   * inner type. From the inner type, we can then add the proper prefix back on.
+   */
+  while (true) {
+    switch (type.GetTypeClass()) {
+      case lldb::eTypeClassArray:
```

Comment:
So I understand, is this case for handling a type name like `struct Foo[]` instead of calling it `Foo[]`? Do we have any tests for this?

Reply:
We don't have a test for the array case. In C, arrays passed as a parameter decay into a pointer.
In C++, it is possible to preserve the array, so I will add a test for it in my next PR.